### ISSUE TITLE
Infinite recursion in IsLayoutDirty

### ISSUE THREAD

**Issue by viciious** [mikke89/RmlUi#41]
Not sure how that happend but here's the stack trace:

<img width="371" alt="изображение" src="https://user-images.githubusercontent.com/1173058/64026058-b36e2480-cb46-11e9-9fa4-0b03bf03aed4.png">

and relevant properties:

<img width="493" alt="изображение" src="https://user-images.githubusercontent.com/1173058/64026144-db5d8800-cb46-11e9-8858-a6a74f1e42a3.png">

**Comment by viciious** [comment]
Might be relevant:
```
RMLUI_ASSERT(!parent || !_parent)
.../libRocket/Source/Core/Element.cpp:1913
```

**Comment by viciious** [comment]
Same thing in Element::GetStyleSheet():

```
const SharedPtr<StyleSheet>& Element::GetStyleSheet() const
{
	if (ElementDocument * document = GetOwnerDocument())
		return document->GetStyleSheet();
```

**Comment by mikke89** [comment]
Hmm, that's strange, thanks for reporting.

Looks like the owner_document is set to an Element that is not an ElementDocument. I'm out traveling now, but will have a closer look at it once I have some more time.

**Comment by viciious** [comment]
Yeah, I think culprit here is that I'm attaching a document to an "iframe" element, which a regular ekement and that worked in libRocket but not in RmlUi.

**Comment by mikke89** [comment]
Ah, yeah, I'm not sure how that would work out. 

So this document under the iframe is an ElementDocument, could the document be replaced by a normal Element? Can I ask what you are using this for? :)

**Comment by viciious** [comment]
I'm using this to create embedded iframes - basically, a document within another document. Each document has its own set of scripts, listeners and styles attached so that makes it impossible to replace them with normal Elements.

By nesting documents you are also able to pass events between them.

**Comment by mikke89** [comment]
Sounds good! I'll have a closer look at it.

**Comment by mikke89** [comment]
I don't see how an Element can be set as an owner_document. Are you setting this manually somehow?

How do you create and attach the document?

**Comment by viciious** [comment]
> 
> 
> I don't see how an Element can be set as an owner_document. Are you setting this manually somehow?
> 
> How do you create and attach the document?

Here's the code (commented out for now):
https://github.com/Qfusion/qfusion/blob/librml/source/ui/widgets/ui_iframe.cpp#L115

**Comment by mikke89** [comment]
I cannot reproduce this issue. I do spot a potential issue with the commented out code though. You are wrapping a non-owning raw pointer in a new smart pointer, which will lead to a double delete at some point (memory corruption), and other potential issues wrt. ownership.

Can you try replacing the commented out code with the following:
```
assert(rocket_document->GetParentNode() != nullptr);
AppendChild( rocket_document->GetParentNode()->RemoveChild(rocket_document) );
```
Assuming the document is still attached to the root (as after loading the document).

There may be other issues lurking though, but I cannot see the owner document ever getting changed in an ElementDocument. In my own testing I'm getting a different assertion failure, not sure if it is serious but I will investigate further.

**Comment by viciious** [comment]
Doh, you're right - that was indeed an oversight of mine. Changing the code to what you suggested fixed the crash, however I can't get the child document to get positioned properly: its position is always seems to be absolute no matter what, even after tweaking the ElementDocument() constructor from:
`SetProperty(PropertyId::Position, Property(Style::Position::Absolute));`
to 
`SetProperty(PropertyId::Position, Property(Style::Position::Relative));`

**Comment by viciious** [comment]
Finally decided to try the Debugger plugin and I must say that I'm pretty damn impressed. Here's what it has to say about the inlined document and its parent iframe "profile_container":

![image](https://user-images.githubusercontent.com/1173058/64488291-ddae9900-d24e-11e9-8695-07d5e01e4ec2.png)

![image](https://user-images.githubusercontent.com/1173058/64488268-b9eb5300-d24e-11e9-8ba5-fc31f7b47f87.png)

**Comment by viciious** [comment]
Just for reference, here's how it is supposed to actually look:

![image](https://user-images.githubusercontent.com/1173058/64488334-50b80f80-d24f-11e9-87fd-324221e0bd45.png)

**Comment by mikke89** [comment]
Can you try this change:

**Comment by mikke89** [comment]
Forgot to mention, in addition you still have to set the position property manually, or set the parent to `position: relative`. 

Okay, I think I understand the need for manual layouting. The dirty layout is set on the nearest document, instead, it needs to be set on the outer document. I think it's easier to just do it manually rather than work around it inside the core library.

**Comment by viciious** [comment]
Hm, it has always worked for me without manual positioning or dirtying of the parent document. The possible reason being that for the iframe element, i'm doing the following:
```
		SetProperty( "display", "inline-block" );
		SetProperty( "overflow", "auto" );
```

**Comment by viciious** [comment]
But yes, that commit did resolve the issue for me, thanks!

**Comment by mikke89** [comment]
Good to hear, no problem!

I think the best solution for iframe is to create a custom IframeElement, which is just a simple replaced element. Then, have the ElementDocument attached to the Context as normal, and draw it in the position of the iframe. I'd rather spend my time on other priorities now though, but something to consider if we want to support `iframe` natively.

**Comment by viciious** [comment]
> Good to hear, no problem!
> 
> I think the best solution for iframe is to create a custom IframeElement, which is just a simple replaced element. Then, have the ElementDocument attached to the Context as normal, and draw it in the position of the iframe. I'd rather spend my time on other priorities now though, but something to consider if we want to support `iframe` natively.

If I'm getting this right, it isn't quite the same thing, as you'll end up with missing scrollbars in case the inline document overflows its parent iframe.

**Comment by mikke89** [comment]
It would work like the html iframe element, we would need width and height defined on the iframe, which we would also use on the iframe body. And probably `overflow: auto` by default, which should give the same behavior?

We wouldn't be able to have the iframe scale dynamically based on its content no, if this is the behavior you need. That is, `width: auto` and `height: auto` wouldn't make sense.

**Comment by viciious** [comment]
No-no, that's not what I mean. The way it currently works for me is exactly the way the the html iframe works. And ye, it has fixed width and height set in css, and defaults to `overflow: auto` and `display: inline-block` in the code, which in turn causes the attached document to have `position` set to `relative`.

https://github.com/Qfusion/qfusion/blob/librml/source/ui/widgets/ui_iframe.cpp#L37

https://github.com/Qfusion/qfusion/blob/librml/source/ui/widgets/ui_iframe.cpp#L113

The iframe also subscribes to `show` and `hide` events of the parent and repeats the events for the framed document.

**Comment by mikke89** [comment]
Oh I see, then the behavior should be the same with how I described. The difference is that the document would not be a child of the iframe, but the context. Then the iframe (or context itself) would be responsible for setting the position and clipping of the document. This way we would avoid any potential issues with nested documents, and also ensure that inherited properties are not inherited "through" the iframe.

But of course, since it works now, I guess no need to spend time on this :)

**Comment by viciious** [comment]
Yes, the way this works right now is alright by me :)

**Comment by mikke89** [comment]
Then it's all good :)

**Comment by viciious** [comment]
Yup, just don't forget to merge the iframe branch into develop ;)
